Refactor Binary Identity Simplification Macro

jmesyou opened the pull request:
This commit cleans up the BINARY_IDENTITY_OP and
BIANRY_IDENTITY_OR_ZERO macros and replaces them
with inline equivalents by implementing a struct to replace template
instantions of the functions tryAndSimplifyBinaryOp*.

This change limits the number of template instantiations
potential call sites have to create as well as abstracts
some captured variables inside the struct to minimize
the number of arguments passed to actual calls.

A closure is emulated by the struct.
Through escape analysis, scalar replacement and inlining,
the overhead of creating this struct should be minimal
on -O3

Related: https://github.com/eclipse/omr/issues/7166

Signed-off-by: James You <james.you@protonmail.com>

hzongaro commented:
It looks like there are some build errors on [Windows](https://github.com/eclipse/omr/pull/7195/checks?check_run_id=19569632683) and [macOS](https://github.com/eclipse/omr/pull/7195/checks?check_run_id=19569617012).

```
D:\a\1\s\compiler\optimizer\OMRSimplifierHandlers.cpp(6550,24): error C2672: 'tryAndSimplifyBinaryIdentityOp': no matching overloaded function found [D:\a\1\s\build\jitbuilder\jitbuilder.vcxproj]
D:\a\1\s\compiler\optimizer\OMRSimplifierHandlers.cpp(6550,7): error C2782: 'TR::Node *tryAndSimplifyBinaryIdentityOp(TR::Simplifier *,T (__cdecl *)(TR::Node *),TR::Node *,T)': template parameter 'T' is ambiguous [D:\a\1\s\build\jitbuilder\jitbuilder.vcxproj]
D:\a\1\s\compiler\optimizer\OMRSimplifierHandlers.cpp(249): message : see declaration of 'tryAndSimplifyBinaryIdentityOp' [D:\a\1\s\build\jitbuilder\jitbuilder.vcxproj]
```

jmesyou commented:
@hzongaro ready for another look!

jmesyou commented on compiler/optimizer/OMRSimplifierHandlers.cpp:
```diff
@@ -137,32 +137,168 @@
+   * If the second child is a constant that represents a zero operation,
+   * replace this node with the second child.
+   */
+   inline TR::Node * tryAndSimplifyIdentityOrZeroOp(TR::Block * block, TR::Node * node, T nullValue, T zeroValue)
```
delete, unused, superseded by BinaryOpSimplifier.tryAndSimplifyIdentityOp

jmesyou commented on compiler/optimizer/OMRSimplifierHandlers.cpp:
```diff
@@ -137,32 +137,168 @@
       }
+};
+
+BinaryOpSimplifier<int8_t> getByteBinaryOpSimplifier(TR::Simplifier * s) {
```
format

jmesyou commented on compiler/optimizer/OMRSimplifierHandlers.cpp:
```diff
@@ -137,32 +137,168 @@
+
+      return NULL;
       }
+};
```
indent

hzongaro commented:
@a7ehuo, I haven't managed to spend the time needed to review this pull request properly.  May I ask you to review?

jmesyou commented on compiler/optimizer/OMRSimplifierHandlers.cpp:
```diff
@@ -137,32 +137,162 @@
+      if (value != zeroValue)
+         return NULL;
+
+      if (performTransformation(s->comp(), "%sFound op with iconst in node [" POINTER_PRINTF_FORMAT "]\n", s->optDetailString(),node))\
```
Just realized there's a dangling `\` here!

a7ehuo commented:
@jmesyou Does the second commit almost rewrite the core change made in the first commit? If so, it might be hard to review since any comment added to the first commit might not be applicable any more after the second commit. Would it make more sense to squash the two commits into one for review? Thank you!

Could the commit comment and the PR description be updated to reflect more accurately on the content since this PR doesn't solve the whole issue but it addresses 2 of the 3 macros mentioned in the issue #7166

```
Fixes: #7166
=>
Related: #7166
```

jmesyou commented:
@a7ehuo I squashed the commits together. Hopefully that will make the PR and changes easier to read. I've also updated the `Fixes` messaged to the `Related` message.

a7ehuo commented on compiler/optimizer/OMRSimplifierHandlers.cpp:
```diff
@@ -6020,8 +6150,11 @@ TR::Node *iaddSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
       {
       return node;
       }
+   auto binOpSimplifier = getIntBinaryOpSimplifier(s);
```
Could an empty line be added before the change for better readability and being consistent with the original code format which has an empty line before `BINARY_IDENTITY_OP`?

 The same comment applies to other places where the new APIs are called.

a7ehuo commented on compiler/optimizer/OMRSimplifierHandlers.cpp:
```diff
@@ -6744,11 +6873,13 @@ TR::Node *daddSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
    // Pretend to be an int before doing the comparison
    //
-   BINARY_IDENTITY_OP(LongInt, DOUBLE_NEG_ZERO)
+   auto binOpSimplifier = getDoubleBinaryOpSimplifier(s);
```
The original code calls macro `BINARY_IDENTITY_OP(LongInt, DOUBLE_NEG_ZERO)` which calls `secondChild->getLongIntType()`. Should the corresponding new API be `getLongBinaryOpSimplifier`?

a7ehuo replied:
Looks like the comment above the original code explains the reasoning: 

https://github.com/eclipse/omr/blob/f010ed6765cfcf7abc64c34b4097093b9ad4fda3/compiler/optimizer/OMRSimplifierHandlers.cpp#L6743-L6750

a7ehuo commented on compiler/optimizer/OMRSimplifierHandlers.cpp:
```diff
@@ -7847,11 +7982,13 @@ TR::Node *dsubSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
    // Pretend to be an int before doing the comparison
    //
-   BINARY_IDENTITY_OP(LongInt, DOUBLE_POS_ZERO)
+   auto binOpSimplifier = getDoubleBinaryOpSimplifier(s);
```
The original code calls macro `BINARY_IDENTITY_OP(LongInt, DOUBLE_POS_ZERO)` which calls `secondChild->getLongIntType()`. Should the corresponding new API be `getLongBinaryOpSimplifier`?

a7ehuo commented on compiler/optimizer/OMRSimplifierHandlers.cpp:
```diff
@@ -10272,8 +10477,11 @@ TR::Node *iushrSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s
    normalizeConstantShiftAmount(node, INT_SHIFT_MASK, secondChild, s);
-   BINARY_IDENTITY_OP(Int, 0)
 
+   auto binOpSimplifier = getShortBinaryOpSimplifier(s);
```
The original code calls `BINARY_IDENTITY_OP(Int, 0)` which calls `secondChild->getIntType()`. Should the corresponding new API be `getIntBinaryOpSimplifier`?

a7ehuo reviewed (commented):
Thank you very much for making this change to clean up the code and squashing the two commits! I just have some minor comments. 

Meanwhile could the description of this PR be updated with the latest commit message? I'd also recommend running some regression tests such as sanity.functional, extended.functional, or sanity.system, sanity.openjdk etc. on different platforms with this change to make sure there is no regression.

jmesyou commented on compiler/optimizer/OMRSimplifierHandlers.cpp:
```diff
@@ -137,32 +137,162 @@
+   * If the second child is a constant that represents an identity operation,
+   * replace this node with the first child.
+   */
+   inline TR::Node * tryAndSimplifyIdentityOp(TR::Node * node, T nullValue)
```
I think it makes more sense to the pass the parent node in (`node`) since that's the binary op we are trying to simplify. It has better encapsulation since you can't accidentally pass in the wrong combination of `node, firstChild, secondChild`, the code will always appropriately get the first and second child from the parent.

jmesyou commented on compiler/optimizer/OMRSimplifierHandlers.cpp:
```diff
@@ -137,32 +137,162 @@
+inline bool isNodeLoadConst(TR::Node * node)
+   {
+   return node != NULL && node->getOpCode().isLoadConst();
+   }
```
I ended up leaving the get/set methods outside of the struct because it is a template. So it didn't make sense to me after I made the change and I could do something like `BinaryOpSimplifier<int32_t>::getNodeFloatBits`. But other than this, I've gone ahead and made all the other changes, please let me know what you think.

a7ehuo commented on compiler/optimizer/OMRSimplifierHandlers.cpp:
```diff
@@ -8532,7 +8680,10 @@ TR::Node *dmulSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
       //
-      BINARY_IDENTITY_OP(LongInt, DOUBLE_ONE)
+      auto binOpSimplifier = getLongBinaryOpSimplifier(s);
+      auto identity = binOpSimplifier.tryAndSimplifyIdentityOp(node, (int64_t) DOUBLE_ONE);
```
I think to be consistent with the rest of the change from this PR on double where `getDoubleBitsBinaryOpSimplifier` is used if the children are double (`dmulSimplifier`), this one should be updated to `getDoubleBitsBinaryOpSimplifier` instead of using `getLongBinaryOpSimplifier`. 

Again since this changes the original code instead of just refactoring, as mentioned earlier, some regression test should be done.

a7ehuo commented on compiler/optimizer/OMRSimplifierHandlers.cpp:
```diff
@@ -9258,7 +9420,11 @@ TR::Node *ddivSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
    //
-   BINARY_IDENTITY_OP(LongInt, DOUBLE_ONE)
+   auto binOpSimplifier = getLongBinaryOpSimplifier(s);
+   auto identity = binOpSimplifier.tryAndSimplifyIdentityOp(node, (int64_t) DOUBLE_ONE);
```
Similar comment as above, for `ddivSimplifier`, I think this one should be updated to `getDoubleBitsBinaryOpSimplifier` instead of using `getLongBinaryOpSimplifier`.

a7ehuo reviewed (commented):
Thank you very much for making the update @jmesyou! I just have some minor follow up comments.

a7ehuo commented on compiler/optimizer/OMRSimplifierHandlers.cpp:
```diff
@@ -8663,7 +8666,7 @@ TR::Node *dmulSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
       // Pretend to be an int before doing the comparison
-      auto binOpSimplifier = getLongBinaryOpSimplifier(s);
+      auto binOpSimplifier = getDoubleBitsBinaryOpSimplifier(s);
       auto identity = binOpSimplifier.tryAndSimplifyIdentityOp(node, (int64_t) DOUBLE_ONE);
```
I don't think `DOUBLE_ONE` should be cast to `int64_t` any more

jmesyou replied:
Good point :eyes:

a7ehuo commented on compiler/optimizer/OMRSimplifierHandlers.cpp:
```diff
@@ -9402,7 +9405,7 @@ TR::Node *ddivSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
    //
-   auto binOpSimplifier = getLongBinaryOpSimplifier(s);
+   auto binOpSimplifier = getDoubleBitsBinaryOpSimplifier(s);
    auto identity = binOpSimplifier.tryAndSimplifyIdentityOp(node, (int64_t) DOUBLE_ONE);
```
I don't think `DOUBLE_ONE` should be cast to `int64_t` any more

a7ehuo reviewed (commented):
Thanks for making the update @jmesyou! I have two comments as below

a7ehuo reviewed (commented):
Thank you for the update! 

LGTM. I assume you're running some regression tests (sanity.functional, extended.functional, sanity.openjdk etc). Once you confirm the regression tests complete successfully and all commits are squashed into one. I'll approve the PR. Thanks again!

jmesyou commented:
Internal build [21052](https://hyc-runtimes-jenkins.swg-devops.com/job/Pipeline-Build-Test-Personal/21052/) is passing

a7ehuo reviewed (approved):
LGTM . Thank you very much for making the update @jmesyou!

hzongaro commented on compiler/optimizer/OMRSimplifierHandlers.cpp:
```diff
@@ -137,32 +137,160 @@
+   /**
+   * Binary identity operation
+   *
+   * If the second child is a constant that represents an identity operation,
```
I know this is copied from the comment that originally accompanied `BINARY_IDENTITY_OP`, but as long as you're here, may I suggest changing "an identity operation" to "an identity operand for the operation", or something like that?

hzongaro commented on compiler/optimizer/OMRSimplifierHandlers.cpp:
```diff
@@ -137,32 +137,160 @@
+   * If the second child is a constant that represents an identity operation,
+   * replace this node with the first child.
+   */
+   inline TR::Node * tryAndSimplifyIdentityOp(TR::Node * node, T nullValue)
```
May I ask you to rename this method to `tryToSimplifyIdentityOp`?

hzongaro replied:
I know the parameter had this name in the original version, but may I suggest renaming `nullValue` to `identityValue` for clarity?

hzongaro commented on compiler/optimizer/OMRSimplifierHandlers.cpp:
```diff
@@ -137,32 +137,160 @@
+   * If the second child is a constant that represents a zero operation,
+   * replace this node with the second child.
+   */
+   inline TR::Node * tryAndSimplifyIdentityOrZeroOp(TR::Block * block, TR::Node * node, T nullValue, T zeroValue)
```
May I ask you to rename this method to `tryToSimplifyIdentityOrZeroOp`?

hzongaro replied:
Again, I know the parameter had this name in the original version, but may I suggest renaming `nullValue` to `identityValue` for clarity?

hzongaro commented on compiler/optimizer/OMRSimplifierHandlers.cpp:
```diff
@@ -137,32 +137,160 @@
+      if (value != zeroValue)
+         return NULL;
+
+      if (performTransformation(s->comp(), "%sFound op with iconst in node [" POINTER_PRINTF_FORMAT "]\n", s->optDetailString(),node))
```
Sorry - I have so many comments on code that wasn't changed.  In this case, the operand might not be an `iconst`.  Perhaps you could change this to use `node->getOpCode().getName()` to get the actual name of the opcode?

hzongaro reviewed (commented):
I think the changes look good.  I have a number of cosmetic comments, mostly on the existing code.

hzongaro commented on compiler/optimizer/OMRSimplifierHandlers.cpp:
```diff
@@ -137,33 +137,161 @@
+      if (value != zeroValue)
+         return NULL;
+
+      auto transformMessage = "%sFound op with %s in node [" POINTER_PRINTF_FORMAT "]\n";
+      if (performTransformation(s->comp(), transformMessage, s->optDetailString(), node->getOpCode().getName(), node))
```
This might seem picky, but I think most &mdash; if not all &mdash; calls to `performTransformation` have the message format string appear directly as an argument to the method.  It can make it easier to find the particular call to `performTransformation` with a regular expression, like `"performTransformation.*Found op"`

jmesyou replied:
squashed (-:

hzongaro reviewed (commented):
Thanks for the updates.  All your changes look good - just one further comment.  Once addressed, you can go ahead and squash the commits.  Thanks!

hzongaro reviewed (approved):
Looks good.  Thanks!

hzongaro commented:
Jenkins build all

hzongaro commented:
It looks like the [z/OS](https://ci.eclipse.org/omr/job/PullRequest-zos_390-64/4356/) and [AIX](https://ci.eclipse.org/omr/job/PullRequest-aix_ppc-64/3993/) compilers are having trouble with [part of these changes](https://github.com/eclipse/omr/blob/0eb490558f585a345d3674c280b46c0fe98021a3/compiler/optimizer/OMRSimplifierHandlers.cpp#L264-L292).

```
13:34:33  "/home/omr/workspace/Build/compiler/optimizer/OMRSimplifierHandlers.cpp", line 266.38: 1540-0063 (S) The text "s" is unexpected.
13:34:33  "/home/omr/workspace/Build/compiler/optimizer/OMRSimplifierHandlers.cpp", line 271.39: 1540-0063 (S) The text "s" is unexpected.
13:34:33  "/home/omr/workspace/Build/compiler/optimizer/OMRSimplifierHandlers.cpp", line 276.39: 1540-0063 (S) The text "s" is unexpected.
13:34:33  "/home/omr/workspace/Build/compiler/optimizer/OMRSimplifierHandlers.cpp", line 281.39: 1540-0063 (S) The text "s" is unexpected.
13:34:33  "/home/omr/workspace/Build/compiler/optimizer/OMRSimplifierHandlers.cpp", line 286.40: 1540-0063 (S) The text "s" is unexpected.
13:34:33  "/home/omr/workspace/Build/compiler/optimizer/OMRSimplifierHandlers.cpp", line 291.40: 1540-0063 (S) The text "s" is unexpected.
```

jmesyou commented:
@hzongaro had to rebase and force push to run the Jenkins pipeline, sorry about that 😅

hzongaro commented:
Jenkins build all

jmesyou commented:
RISV failure looks unrelated, had similar failure yesterday on unrelated PR **https://github.com/eclipse/omr/pull/7287**. MacOS failure known

hzongaro commented:
> RISV failure looks unrelated, had similar failure yesterday on unrelated PR https://github.com/eclipse/omr/pull/7287

Opened issue #7291 for the RISC-V failure.
